### PR TITLE
Mix jmap imap luc

### PR DESCRIPTION
**Pull request by chibenwa**
Two last commits to review

### CODE REVIEW

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 1]
```diff
@@ -51,6 +51,17 @@ public void hasMessageInMailbox(String mailbox) throws Throwable {
             .isTrue();
     }
 
+    @Then("^the user has a IMAP notification about (\\d+) new message on mailbox \"([^\"]*)\"$")
+    public void hasANotificationAboutNewMessagesInMailbox(int numOfNewMessage, String mailbox) throws Throwable {
+        IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);
```
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Should we naming for host and port value?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature [hunk 2]
```diff
@@ -19,4 +19,9 @@ Feature: IMAP compatibility of JMAP setMessages method used to update mailboxes
     Given the user has a message "m1" in "source" mailbox with subject "My awesome subject", content "This is the content"
     When the user copy "m1" from mailbox "source" to mailbox "mailbox"
     Then the user has a IMAP message in mailbox "mailbox"
-    And the user has a IMAP message in mailbox "source"
\ No newline at end of file
+    And the user has a IMAP message in mailbox "source"
+
+  Scenario: If a message is moved byJMAP, IMAP connexion will be notified
```
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature]: s/byJMAP/by JMAP

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 3]
```diff
@@ -33,11 +37,13 @@
 
     private final MainStepdefs mainStepdefs;
     private final UserStepdefs userStepdefs;
+    private final Map<String, IMAPMessageReader> imapConnections;
 
     @Inject
     private ImapStepdefs(MainStepdefs mainStepdefs, UserStepdefs userStepdefs) {
         this.mainStepdefs = mainStepdefs;
         this.userStepdefs = userStepdefs;
+        this.imapConnections = new HashMap<>();
```
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Or can we use Maps.newHashMap()?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 4]
```diff
@@ -51,7 +57,7 @@ public void hasMessageInMailbox(String mailbox) throws Throwable {
             .isTrue();
     }
 
-    @Then("^the user has a IMAP notification about (\\d+) new message on mailbox \"([^\"]*)\"$")
+    @Then("^the user has a IMAP notification about (\\d+) new message when selecting mailbox \"([^\"]*)\"$")
     public void hasANotificationAboutNewMessagesInMailbox(int numOfNewMessage, String mailbox) throws Throwable {
         IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);
```
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Naming for constant values

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 5]
```diff
@@ -72,4 +78,22 @@ public void hasNoMessageInMailbox(String mailbox) throws Throwable {
                 mailbox))
             .isTrue();
     }
+
+    @Given("^the user has a open IMAP connexion with mailbox \"([^\"]*)\" selected")
+    public void openImapConnexionAndSelectMailbox(String mailbox) throws Throwable {
+        IMAPMessageReader imapMessageReader = new IMAPMessageReader("127.0.0.1", 1143);
```
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Naming for constant values

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java [hunk 6]
```diff
@@ -72,4 +78,22 @@ public void hasNoMessageInMailbox(String mailbox) throws Throwable {
                 mailbox))
             .isTrue();
     }
+
+    @Given("^the user has a open IMAP connexion with mailbox \"([^\"]*)\" selected")
+    public void openImapConnexionAndSelectMailbox(String mailbox) throws Throwable {
```
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: "Connexion" is French?

**Comment by chibenwa** [comment]
test this please

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java (hunk 6 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: Seems like

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java [hunk 7]
```diff
@@ -68,6 +85,34 @@ private String readFirstMessageInInbox(String user, String password, String para
...
+        imapClient.noop();
+
+        String replyString = imapClient.getReplyString();
+        List<String> parts = Splitter.on('\n')
+            .trimResults()
+            .omitEmptyStrings()
+            .splitToList(replyString);
+        return parts.size() == 3
+            && parts.get(2).contains("OK NOOP completed.")
+            && parts.contains("* " + uid + " EXISTS")
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java]: That's not a uid

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java [hunk 8]
```diff
@@ -68,6 +85,34 @@ private String readFirstMessageInInbox(String user, String password, String para
...
+            && parts.contains("* " + uid + " RECENT");
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java]: neither

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java [hunk 9]
```diff
@@ -68,6 +85,34 @@ private String readFirstMessageInInbox(String user, String password, String para
...
+
+        String replyString = imapClient.getReplyString();
+        List<String> parts = Splitter.on('\n')
+            .trimResults()
+            .omitEmptyStrings()
+            .splitToList(replyString);
+
+        return parts.size() == 2
+            && parts.get(1).contains("OK NOOP completed.")
+            && parts.contains("* " + uid + " EXPUNGE");
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/IMAPMessageReader.java]: still not an uid I think (should be a msn, no?)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            "    }," +
+            "    \"#0\"" +
+            "  ]" +
+            "]";
+        Request.Post(mainStepdefs.baseUri().setPath("/jmap").build())
+            .addHeader("Authorization", userStepdefs.tokenByUser.get(username).serialize())
+            .bodyString(requestBody, org.apache.http.entity.ContentType.APPLICATION_JSON)
+            .execute()
+            .discardContent();
+        Thread.sleep(1000L);
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/SetMessagesMethodStepdefs.java]: await?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+  Scenario: If a message is moved by JMAP, IMAP client will be notified when selecting mailbox
+    Given the user has a message "m1" in "source" mailbox with subject "My awesome subject", content "This is the content"
+    When the user move "m1" to mailbox "mailbox"
+    Then the user has a IMAP notification about 1 new message when selecting mailbox "mailbox"
+
+  Scenario: If a message is moved by JMAP, IMAP client that have selected the destination mailbox will be notified
+    Given the user has a message "m1" in "source" mailbox with subject "My awesome subject", content "This is the content"
+    Given the user has a open IMAP connexion with mailbox "mailbox" selected
+    When the user move "m1" to mailbox "mailbox"
+    Then the user has a IMAP RECENT and  notification about new message with uid 1 on connexion for mailbox "mailbox"
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature]: 2 spaces before notification? missing word after and?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+  Scenario: If a message is moved by JMAP, IMAP client that have selected the source mailbox will not be notified
+    Given the user has a message "m1" in "source" mailbox with subject "My awesome subject", content "This is the content"
+    Given the user has a open IMAP connexion with mailbox "source" selected
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/ImapSetMessagesMailboxesUpdatesCompatibility.feature]: *an open

**Comment by chibenwa** [comment]
See https://github.com/linagora/james-project/pull/582

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java (hunk 3 above)
Comment (strokyl) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/ImapStepdefs.java]: What's the advantage of Maps.newHashMap() ?

**Pull request closed without merging**